metrics: experimental internal metrics

VDanielEdwards opened the pull request:
Issue: SILKIT-1379

<!--
SPDX-FileCopyrightText: 2023 Vector Informatik GmbH

SPDX-License-Identifier: MIT
-->

## Subject
<!--
    - What kind of changes does this pull request comprise?
    - Add a label like "Bugfix", "docs", etc. to this PR to give context.
 -->

Creation, collection, sending, receiving, and storing various metrics as JSON.

The metrics can be used to get insights into internal behavior of individual SIL Kit participants or simulations.

Currently available metrics:
- `COUNTER`: Collects single unsigned integer value.
- `STATISTICS`: Collects single `double` values and accumulates minimum, maximum, exponential incremental mean, and standard deviation.
- `STRING_LIST`: Collects a list of strings.

If metrics are sent to remote participants, the participant that collects these, should be started first. The registry also supports collection of remote metrics, which is preferable, since it must be started first anyway.

Collection and storage of the metrics is configured via the participant configuration file (or the registry configuration file).

Metrics are only sent if they have changed.

## Description
<!--
    - Give a short summary of the intention of the changes.
    - Don't replicate the commit messages here.
    - Which Jira tickets are addressed with this pull request (list all if multiple tickets are affected)?
-->

## Instructions for review / testing
<!--
    - Hilight some of the important changes, which reviewers should focus on
    - Which parts should be reviewed in detail? For example: content of console output, correct semantics of changes
    - Test steps and test setup description. For example: which programs or configs to use
-->

## Developer checklist (address before review)

- [ ] Changelog.md updated
- [ ] Prepared update for depending repositories
- [ ] Documentation updated (public API changes only)
- [ ] API docstrings updated (public API changes only)
- [ ] Rebase &rarr; commit history clean
- [x] Squash and merge &rarr; proper PR title

VDanielEdwards commented on SilKit/source/core/internal/traits/SilKitMsgTraits.hpp:
```diff
@@ -126,81 +126,86 @@ struct SilKitMsgTraits
     };
 
 DefineSilKitMsgTrait_TypeName(SilKit::Services::Logging, LogMsg) DefineSilKitMsgTrait_TypeName(
```
I'm going to create a separate PR to fix the messy formatting of this file in particular.

VDanielEdwards replied:
The PR exists now here: https://github.com/vectorgrp/sil-kit/pull/86

VDanielEdwards commented on SilKit/source/core/internal/traits/SilKitMsgSerdesName.hpp:
```diff
@@ -46,6 +46,8 @@ struct SilKitMsgTraitSerdesName
     }
 
 DefineSilKitMsgTrait_SerdesName(SilKit::Services::Logging::LogMsg, "LOGMSG");
+DefineSilKitMsgTrait_SerdesName(VSilKit::MetricsRegistration, "METRICSREGISTRATION");
```
The `MetricsRegistration` type isn't used anymore.

```suggestion
```

VDanielEdwards commented:
@KonradBkd @MariusBgm I'll rebase onto main to get the much much cleaner formatting of the traits header, if you don't mind. I'm going to do that if both of you give this comment a thumbs-up reaction.

KonradBkd commented on SilKit/source/services/metrics/MetricsTimerThread.cpp:
```diff
@@ -0,0 +1,78 @@
+
+        while (true)
+        {
+            if (done.wait_for(std::chrono::seconds{1}) != std::future_status::timeout)
```
If I got it right, the 1s here will be the update time for writing / sending the metrics. Don't we want to configure this?

VDanielEdwards replied:
Yes that's correct. In the long run this will be a configuration setting, but for the initial experiment it doesn't have to be.

KonradBkd reviewed (commented):
Review Part 1/?: 
Nice! So far I ran over the services/metrics, services/orchestration parts and just found some minor stuff. Will test it on some Demos now...

VDanielEdwards commented on SilKit/source/services/metrics/MetricsProcessor.cpp:
```diff
@@ -0,0 +1,131 @@
+
+namespace {
+
+auto CurrentTimestamp() -> std::string
```
That is a really good idea. I'll put it in the new `StringHelpers.hpp` / `.cpp` files.

VDanielEdwards commented on SilKit/source/services/metrics/MetricsDatatypes.cpp:
```diff
@@ -0,0 +1,59 @@
+              << ", kind=" << metricData.kind << ", value=" << metricData.value << "}";
+}
+
+auto operator<<(std::ostream& os, const MetricsUpdate& metricsUpdate) -> std::ostream&
```
Thats really clever, I think I'll take that suggestion :+1:

KonradBkd commented on SilKit/source/services/orchestration/TimeSyncService.cpp:
```diff
@@ -475,10 +478,15 @@ void TimeSyncService::ExecuteSimStep(std::chrono::nanoseconds timePoint, std::ch
     _waitTimeMonitor.StopMeasurement();
-    Trace(_logger, "Starting next Simulation Task. Waiting time was: {}ms",
-          std::chrono::duration_cast<DoubleMSecs>(_waitTimeMonitor.CurrentDuration()).count());
+    const auto waitingDuration = _waitTimeMonitor.CurrentDuration();
```
This gets called before the first call to `_waitTimeMonitor.StartMeasurement();`. So the first waitingDuration will be max

VDanielEdwards replied:
The first sample is now skipped. Thanks! :+1:

KonradBkd reviewed (commented):
Does it ever make sense to have a `Remote` sink and `CollectFromRemote: True` on the same participant? I managed to get a kind of recursion with the setup:

P1:
```
Experimental:
  Metrics:
    Sinks:
      - Type: Remote
    CollectFromRemote: True
```

P2:
```
Experimental:
  Metrics:
    Sinks:
      - Type: JsonFile
        Name: PartMetricsLogRemotes
    CollectFromRemote: True
```

VDanielEdwards commented:
> Does it ever make sense to have a `Remote` sink and `CollectFromRemote: True` on the same participant? I managed to get a kind of recursion with the setup:
> 
> P1:
> 
> ```
> Experimental:
>   Metrics:
>     Sinks:
>       - Type: Remote
>     CollectFromRemote: True
> ```
> 
> P2:
> 
> ```
> Experimental:
>   Metrics:
>     Sinks:
>       - Type: JsonFile
>         Name: PartMetricsLogRemotes
>     CollectFromRemote: True
> ```

Fixed and added a test that ensures that such configs are not accepted.

MariusBgm commented on SilKit/source/util/ExecutionEnvironment.cpp:
```diff
@@ -0,0 +1,281 @@
+auto GetPhysicalMemoryMB() -> std::string
+{
+    static const auto result = [] {
+        const auto pageSize = static_cast<std::uint64_t>(::sysconf(_SC_PAGESIZE));
```
sysconf is as posix as it gets. this could be inside a `#if __unix__`, I think

VDanielEdwards replied:
It is inside of the `#ifdef __unix__`, just hard to see. Maybe I should place them around each function individually ...

VDanielEdwards replied:
I restructured the entire file to make it (hopefully) much more readable!

MariusBgm reviewed (commented):
Just some nit-picking comments and reminders for myself. take as low-prio inputs :-)

VDanielEdwards commented on SilKit/source/services/metrics/MetricsManager.cpp:
```diff
@@ -0,0 +1,338 @@
+    const auto increment = alpha * difference;
+
+    _mean += increment;
+    _variance = (1.0 - alpha) * (_variance + difference * increment);
```
I talked to @KonradBkd as well, the parameter should in time be configurable for each individual statistic metric, depending on how often the value is expected to be updated.

VDanielEdwards commented on SilKit/source/core/participant/Participant_impl.hpp:
```diff
@@ -191,6 +213,33 @@ void Participant<SilKitConnectionT>::SetupRemoteLogging()
+    processor.SetSender(*sender);
+    processor.SetupSinks(_participantConfig);
+
+    // NB: Create the metrics manager prior to anything that might need it (possibly needs to be split into
```
This comment is indeed out-of-date. I removed it! The split makes it easier to integrate into both `VAsioRegistry` and `Participant<...>`. It also makes it a bit easier to write unit-tests for the manager.

VDanielEdwards commented on SilKit/source/core/mock/participant/MockParticipant.hpp:
```diff
@@ -176,6 +178,67 @@ class DummyNetworkSimulator : public Experimental::NetworkSimulation::INetworkSi
     void Start() override {}
 };
 
+class MockMetricsManager : public IMetricsManager
```
Indeed, this is class isn't named correctly. Fixed!

VDanielEdwards commented on SilKit/source/services/metrics/MetricsTimerThread.hpp:
```diff
@@ -0,0 +1,36 @@
+class MetricsTimerThread : public IMetricsTimerThread
+{
+    std::atomic<bool> _started{false};
+    std::promise<void> _go;
```
That would be the perfect place for a `std::latch` (or a `std::barrier`) (but they are only available in C++ 20 onwards).

VDanielEdwards commented on SilKit/source/util/ExecutionEnvironment.cpp:
```diff
@@ -0,0 +1,281 @@
+
+auto GetProcessorCount() -> std::string
+{
+    static const auto result = std::to_string(::get_nprocs());
```
`get_nprocs` is a `glibc` specific function. I completely forgot about `hardware_concurrency`, thats much better!

VDanielEdwards commented on SilKit/source/util/ExecutionEnvironment.cpp:
```diff
@@ -0,0 +1,281 @@
+
+auto GetProcessExecutable() -> std::string
+{
+    return UNKNOWN_VALUE;
```
Ah thank you 😸 I hoped you would jump in with a suggestion for non-Linux systems! I'll add it as a comment, but won't actually put it.

VDanielEdwards commented on SilKit/source/core/internal/traits/SilKitMsgSerdesName.hpp:
```diff
@@ -46,6 +46,7 @@ struct SilKitMsgTraitSerdesName
     }
 
 DefineSilKitMsgTrait_SerdesName(SilKit::Services::Logging::LogMsg, "LOGMSG");
+DefineSilKitMsgTrait_SerdesName(VSilKit::MetricsUpdate, "METRICVALUE");
```
This one is not a nit-pick! Good catch! Thank you 😸

KonradBkd reviewed (approved):
All of my comments were addressed, LGTM!

MariusBgm reviewed (approved):
LGTM!
